text/trace/0192-messaging-semantic-conventions-spec.md

```diff
@@ -0,0 +1,443 @@
+The operations modelled by "Deliver" or "Receive" spans do not strictly refer
+to receiving the message from intermediaries, but instead refer to the
+application receiving messages for processing. If messages are fetched from the
+intermediary and forwarded to the application in one go, the whole operation
+might be covered by a "Deliver" or "Receive" span. However, clients might
+pre-fetch messages from intermediaries and cache those messages, and only
+forward messages to the application at a later time. In this case, the
+operation of pre-fetching and caching should not be covered by the "Deliver" or
+"Receive" spans.
+
+Operations covered by "Deliver" or "Receive" can forward zero messages (e. g.
+to notify the application that no message is available for processing), one
+message, or multiple messages (a batch of messages). "Deliver" and "Receive"
+spans should link to the "Create" span of the messages forwarded, thus those
+spans can link to zero, one, or multiple "Create" spans.
+
+> "Deliver" or "Receive" spans MUST NOT be created for messages which are not
+> forwarded to the application, but are pre-fetched or cached by messaging
+> libraries or SDKs.
+>
+> A single "Deliver" or "Receive" span can account for a single message, for
+> multiple messages (in case messages are passed for processing as batches), or
+> for no message at all (if it is signalled that no messages were received).  For
+> each message it accounts for, the "Deliver" or "Receive" span SHOULD link to
+> the "Create" span for the message.
+
+##### Settlement
+
+Messages can be settled in a variety of different ways. In some cases, messages
+are not settled at all (fire-and-forget), in other cases settlement operations
+are triggered manually by the user, in callback scenarios settlement can be
+automatically triggered by messaging SDKs based on return values of callbacks.
+
+A "Settle" spans should be created for every settlement operation, no matter if
+the settlement operation was manually triggered by the user or automatically
+triggered by SDKs. SDKs will in some cases auto-settle messages in
+push-scenarios, when messages are delivered via callbacks. In this case it is
+recommended to create a parent span, so that a "Settle" span will be a sibling
+of the related "Deliver" span.
+
+In cases where it is not feasible to create spans for settlement operations, an
+event should be created instead. Events could be added to "Deliver" spans or to
+ambient spans.
+
+"Settle" spans may link to "Create" spans of the messages that are settled,
+however, for some settlement scenarios this is not feasible or possible.
+
+> "Settle" spans or events SHOULD be created for every manually or automatically
+> triggered settlement operation. A single "Settle" span can account for a
+> single message or for multiple messages (in case messages are passed for
+> settling as batches). For each message it accounts for, the "Settle" span
+> MAY link to the "Create" span for the message.
```

Comment:
When we discussed the "deliver" span, we decided to use the links array as a container to specify **all** messages involved in the batch. This way we can:
- enumerate the participating messages in a consistent and stable manner
- have a silo to store "per message" attributes, like the id of the message being settled

I think we can benefit from using this pattern for settle as well, which will be identical semantics for the two similar operations.

Reply:
Are you talking about link attributes, or about an array of message ids in a single attribute? We didn't add this for deliver spans in this document. We could discuss this.

I tried to keep the requirements for "Settle" as close to "Deliver" as possible:

> "Deliver" or "Receive" spans MUST NOT be created for messages which are not forwarded to the application, but are pre-fetched or cached by messaging libraries or SDKs.
>
> A single "Deliver" or "Receive" span can account for a single message, for multiple messages (in case messages are passed for processing as batches), or for no message at all (if it is signalled that no messages were received).  For each message it accounts for, the "Deliver" or "Receive" span SHOULD link to the "Create" span for the message.